src/esp/geo/VoxelGrid.h

```diff
@@ -0,0 +1,515 @@
+   * directory, all .svx files will be loaded. If file path points to a file,
+   * this file must represent a boolean voxel grid which will represent the
+   * voxel grid's Boundary grid.
+   * @param filepath Can either be the path to a directory or a single .svx
+   * file.
+   */
+  explicit VoxelGrid(const std::string filepath);
+
+  /**
+   * @brief Generates a new empty voxel grid of a specified type.
+   * @param name The key underwhich the grid will be registered and accessed.
+   */
+  template <typename T>
+  void addGrid(const std::string& gridName) {
+    std::string type;
+    // set type depending on the type of T
+    if (std::string(typeid(T).name()) == "i") {
+      type = "int";
+    } else if (std::string(typeid(T).name()) == "f") {
+      type = "float";
+    } else if (std::string(typeid(T).name()) == "b") {
+      type = "bool";
+    } else if (std::string(typeid(T).name()) == "N6Magnum4Math7Vector3IfEE") {
+      type = "vector3";
+    }
+
+    if (grids_.find(gridName) != grids_.end()) {
+      // grid exists, simply overwrite
+      Mn::Debug() << gridName << "exists, overwriting.";
+      Corrade::Containers::Array<char> cr_grid{Corrade::Containers::ValueInit,
+                                               gridSize() * sizeof(T)};
+      grids_[gridName].second = std::move(cr_grid);
+      grids_[gridName].first = type;
+      return;
+    }
+    Corrade::Containers::Array<char> cr_grid{Corrade::Containers::ValueInit,
+                                             gridSize() * sizeof(T)};
+    grids_.insert(
+        std::make_pair(gridName, std::make_pair(type, std::move(cr_grid))));
+  }
+
+  /**
+   * @brief Removes a grid and frees up memory.
+   * @param name The name of the grid to be removed.
+   */
+  void removeGrid(const std::string& gridName) {
+    assert(grids_.find(gridName) != grids_.end());
+    grids_.erase(gridName);
+  }
+
+  template <typename T>
+  Cr::Containers::StridedArrayView<3, T> getGrid(const std::string& gridName) {
+    std::string type = grids_[gridName].first;
+    unsigned long dims[3]{static_cast<unsigned long>(m_voxelGridDimensions[0]),
+                          static_cast<unsigned long>(m_voxelGridDimensions[1]),
+                          static_cast<unsigned long>(m_voxelGridDimensions[2])};
+    return Cr::Containers::StridedArrayView<3, T>{
+        Cr::Containers::arrayCast<T>(grids_[gridName].second), dims};
+  }
+
+  /**
+   * @brief Linearizes 3D voxel coordinates to a single value in order to
+   * directly access a voxel grid.
+   * @param coords The 3D voxel index to be linearized.
+   * @return The has value of the voxel index.
+   */
+  int hashVoxelIndex(const Mn::Vector3i& coords);
+
+  /**
+   * @brief Converts a hash value into a 3 dimensional coordinate.
+   * @param hash The hash value.
+   * @return The voxel index derived from the given hash value.
+   */
+  Mn::Vector3i reverseHash(const int hash);
```

Comment:
As with `setVoxelByHash()` / `getVoxelByHash()`, I don't see a reason for these to exist anymore :)